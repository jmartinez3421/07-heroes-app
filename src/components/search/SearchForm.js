import React from 'react';
import { useSimpleForm } from '../../hooks/useSimpleForm'

export const SearchForm = ({history}) => {

    const [hero, handleInputChange] = useSimpleForm({name: ''})

    const handleSubmit = (e) => {
        e.preventDefault();
        
        history.push(`?q=${hero.name}`);
    };

    return (
        <>
            <h4>Search your Hero</h4>
            <hr />

            <form onSubmit={handleSubmit}>

                <input
                    type='text'
                    placeholder='Search a hero'
                    className='form-control'
                    value={hero.name}
                    name='name'
                    onChange={e => handleInputChange(e)}
                    autoComplete='off'
                />

                <div className="d-grid gap-2 mt-1">
                    <button
                        className="btn btn-outline-success"
                        type="submit"
                        onSubmit={handleSubmit}
                    >
                        Search
                    </button>
                </div>

            </form>
        </>
    )
}
