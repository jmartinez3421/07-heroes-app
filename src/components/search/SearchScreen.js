import React, { useMemo } from 'react'
import queryString from 'query-string';

import { useLocation } from 'react-router';
import { getHeroByName } from '../../selectors/getHeroeByName';
import { HeroeCard } from '../heroes/HeroeCard';
import { SearchForm } from './SearchForm'

export const SearchScreen = ({history}) => {
    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);

    const heroesFiltered = useMemo(() => getHeroByName(q), [q]);

    return (
        <div>
            <div className='row'>

                <div className='col-5'>
                    <SearchForm history={history} />
                </div>

                <div className='col-7'>
                    <h4>Results</h4>
                    <hr />

                    {
                        (q === '') &&
                            <div className='alert alert-info'>
                                Search a hero...
                            </div>
                    }

                    {
                        (q !== '' && heroesFiltered.length === 0) &&
                            <div className='alert alert-danger'>
                                There is no hero with: {q}
                            </div>
                    }

                    {
                    
                        heroesFiltered.map(hero => (
                            <HeroeCard key={hero.id}  heroe={hero} />
                        ))

                    }
                </div>

            </div>
        </div>
    )
}
