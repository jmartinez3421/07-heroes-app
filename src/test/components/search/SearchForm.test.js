import { mount } from 'enzyme';
import React from 'react';
import { MemoryRouter, Route } from 'react-router';

import {SearchForm} from '../../../components/search/SearchForm';


describe('Pruebas sobre el componente <SearchForm />', () => {
    const history = {
        push: jest.fn()
    }

    test('Debe de mostrarse correctamente con los valores por defecto', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path='/search' component={() => <SearchForm history={history} />}/>  
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('');
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de llamar el push del history', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path='/search' component={() => <SearchForm history={history} />}/>  
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change',{
            target: {
                name: 'name',
                value: 'batman'
            }
        })

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        })

        expect(history.push).toHaveBeenCalledWith('?q=batman');
    })
    
    
})
