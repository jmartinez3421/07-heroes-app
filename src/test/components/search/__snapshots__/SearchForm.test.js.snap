// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre el componente <SearchForm /> Debe de mostrarse correctamente con los valores por defecto 1`] = `
Array [
  <h4>
    Search your Hero
  </h4>,
  <hr />,
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="name"
      onChange={[Function]}
      placeholder="Search a hero"
      type="text"
      value=""
    />
    <div
      className="d-grid gap-2 mt-1"
    >
      <button
        className="btn btn-outline-success"
        onSubmit={[Function]}
        type="submit"
      >
        Search
      </button>
    </div>
  </form>,
]
`;
